File: 2015/04/04.cpp
#include <iostream>
#include <string>
#include "./md5/md5.h"

int main (void) {

    std::string problem_input = "ckczppom";
    std::string string_key, result;
    
    bool run_once = true;
    for (unsigned int i = 0; i < 10'000'000; i++) {

        string_key = problem_input + std::to_string(i);
        result = md5(string_key);

        // Pt. 1
        if (run_once && result.substr(0, 5) == "00000") {
            run_once = false;
            std::cout << i << std::endl;
        }

        // Pt. 2
        if (result.substr(0, 6) == "000000") {
            std::cout << i << std::endl;
            break;
        }
    }

}

// 3489 ms ± 584 ms (20 runs, 10 warmup) - Clang 14.0.6: clang++ -Ofast -std=C++20

File: 2015/04/04.rs
// rust-crypto = "^0.2"

use crypto::digest::Digest;

fn main() {

    let problem_input = "ckczppom";
    let mut run_once = true;
    let mut hasher = crypto::md5::Md5::new();

    for i in 0..10_000_000 {
        let string_key = problem_input.to_string() + &i.to_string();
        hasher.input(string_key.as_bytes());
        let result = hasher.result_str();
        hasher.reset();

        // Pt. 1
        if run_once && &result[..5] == "00000" {
            run_once = false;
            println!("{i}");
        }

        // Pt. 2
        if &result[..6] == "000000" {
            println!("{i}");
            return;
        }

    }

}

// 837.3 ms ± 19.1 ms (50 runs, 25 warmup) - Rustc 1.66.0: -opt-level=3
